iris_grib/tests/unit/load_convert/test_product_definition_section.py

```diff
@@ -0,0 +1,145 @@
+            'typeOfSecondFixedSurface'
+        ]
+
+    def _check_fixed_surface(self, fs_is_expected, fs_is_present):
+        """
+        Whether or not fixed surface elements are expected/present in the
+        section 4 keys, most of the code is shared so we are using a single
+        function with parameters.
+        """
+
+        string_expected = 'Expected' if fs_is_expected else 'Unexpected'
+        string_present = 'Present' if fs_is_present else 'Absent'
+        print('Fixed Surface Keys *{}* and *{}*... '.format(string_expected,
+                                                            string_present))
+
+        # Use the section 4 from either product_definition_section #1 or #31.
+        # #0 contains fixed surface elements, #31 does not.
+        template_number = 0 if fs_is_expected else 31
+        section_4 = self.templates[template_number]
+        section_4.update({
+            'productDefinitionTemplateNumber': template_number,
+            'parameterCategory': None,
+            'parameterNumber': None
+        })
+
+        for key in self.fixed_surface_keys:
+            # Force the presence or absence of the fixed surface elements even
+            # when they're respectively ignored or expected.
+            if fs_is_present and key not in section_4:
+                section_4[key] = pdt_0_section_4()[key]
+            elif (not fs_is_present) and key in section_4:
+                del section_4[key]
+
+        def run_function():
+            # For re-use in every type of test below.
+            product_definition_section(
+                section_4, self.metadata, self.discipline, self.tablesVersion,
+                self.rt_coord)
+
+        if fs_is_expected and not fs_is_present:
+            # Should error since the expected keys are missing.
+            error_message = 'FixedSurface'
+            with six.assertRaisesRegex(self, KeyError, error_message):
+                run_function()
+        else:
+            # Should have a successful run for all other circumstances.
+
+            # Translate_phenomenon_patch is the end of the function,
+            # and should be able to accept None for the fixed surface
+            # arguments. So should always have run.
+            previous_call_count = self.translate_phenomenon_patch.call_count
+            run_function()
+            self.assertEqual(self.translate_phenomenon_patch.call_count,
+                             previous_call_count + 1)
+            phenom_call_args = self.translate_phenomenon_patch.call_args[1]
+            for key in self.fixed_surface_keys:
+                # Check whether None or actual values have been passed for
+                # the fixed surface arguments.
+                if fs_is_expected:
+                    self.assertEqual(phenom_call_args[key], section_4[key])
+                else:
+                    self.assertIsNone(phenom_call_args[key])
+
+        print('passed')
```

Comment:
@trexfeathers I see the intent behind this, but there is no guarantee that the `print` output will make sense for a testing framework that is running tests in parallel... would you mind removing this?

Reply:
👍